### ISSUE TITLE
java.lang.ArrayIndexOutOfBoundsException (negative cluster) for subqueries

### ISSUE THREAD

**Issue by benjaminbillet** [orientechnologies/orientdb#6735]
## OrientDB Version, operating system, or hardware.
v2.2.10, Mac OS X

## Expected behavior and actual behavior
An ArrayIndexOutOfBoundsException is thrown when my java client tries to access "intermediary" vertices created by a subquery, e.g., `SELECT $a LET $a = (SELECT @rid AS parentId, out('AtoB') AS vertices FROM TypeA) FETCHPLAN $a:2`

The result in studio looks good, but the java client crashes while iterating on the property $a.
 
```
$a
-----
[{"@type":"d","@rid":"#-2:0","@version":0,"parentId":{"@type":"d","@rid":"#171:6","@version":3,"@class":"TypeA","out_AtoB":["#189:12","#190:12"],"@fieldTypes":"out_AtoB=g"},"vertices":[{"@type":"d","@rid":"#181:12","@version":2,"@class":"TypeB","in_AtoB":["#189:12"],"@fieldTypes":"in_AtoB=g"},{"@type":"d","@rid":"#182:12","@version":2,"@class":"TypeB","in_AtoB":["#190:12"],"@fieldTypes":"in_AtoB=g"}],"@fieldTypes":"vertices=z"}] 
```

The java code:
```java
OrientGraph graph = ...;

OrientVertex root = graph.addVertex("class:TypeA");
graph.commit();

for (int i = 0; i < 2; i++) {
    OrientVertex v = graph.addVertex("class:TypeB");
    root.addEdge("AtoB", v);
}
graph.commit();

String query = "SELECT $a LET $a = (SELECT @rid AS parentId, out('AtoB') AS vertices FROM TypeA) FETCHPLAN $a:2";

Iterable<OrientVertex> results = graph.command(new OCommandSQL(query)).execute();
OrientVertex result = results.iterator().next();

Iterable<OrientVertex> vertices = result.getProperty("$a"); // com.tinkerpop.blueprints.impls.orient.OrientElementIterable
for (OrientVertex vertex : vertices) { // ArrayIndexOutOfBoundsException when next is called
}
```

```
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: -2
        at com.orientechnologies.orient.client.remote.OStorageRemote.getPhysicalClusterNameById(OStorageRemote.java:1479)
        at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.executeReadRecord(ODatabaseDocumentTx.java:1890)
        at com.orientechnologies.orient.core.tx.OTransactionNoTx.loadRecord(OTransactionNoTx.java:112)
        at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.load(ODatabaseDocumentTx.java:1592)
        at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.load(ODatabaseDocumentTx.java:101)
        at com.orientechnologies.orient.core.id.ORecordId.getRecord(ORecordId.java:322)
        at com.tinkerpop.blueprints.impls.orient.OrientElementIterator.next(OrientElementIterator.java:62)
        at com.tinkerpop.blueprints.impls.orient.OrientElementIterator.next(OrientElementIterator.java:35)
        at Test5.main(Test5.java:42)
```

Additional informations:
- The problem occurs with and without a FETCHPLAN clause.
- The query is a demo for the problem. n my actual queries, I have several variables in the result set. Consequently, as far as i know, I can't use expand.

Intuitively, I think that the problem is related to the "virtual" vertices (-2:0 and -2:1) but I don't know why the java client tries to reload these vertices. Maybe am I using something in the wrong way?

PS: I'm not sure that the current title is clear enough for the problem.

**Comment by tglman** [comment]
hi all,

just pushed some fixes about this, they are going to be released in 2.2.11.

Closing this

**Comment by benjaminbillet** [comment]
Thanks a lot. I can't wait for this 2.2.11, it will solve a lot of my issues 👍
